practical/ProcessMining/group2/inductive_miner/src/graph_utils.py

```diff
@@ -0,0 +1,270 @@
+                sccs.append(current_scc)
+
+        return sccs
+
+    def build_cuts_graph(self, cuts: List[List[str]]) -> Tuple['Graph', Dict[str, int]]:
+        """
+        Builds a graph representing connections between cuts and maps each node to its corresponding cut index.
+        :param cuts: List of lists, where each inner list represents a cut.
+        :return: Tuple containing the resulting graph (always a path) and a mapping of nodes to their cut indices.
+        """
+        cut_graph = {i: set() for i in range(len(cuts))}
+        cut_map = {node: i for i, scc in enumerate(cuts) for node in scc}
+
+        for node in self.get_all_nodes():
+            current_cut = cut_map[node]
+            neighbors = set(self.get_neighbors(node))
+            
+            for neighbor in neighbors:
+                neighbor_cut = cut_map[neighbor]
+                # If the neighbor belongs to a different cut, add its cut index to the current cut's set
+                if current_cut == neighbor_cut:
+                    continue
+                if len(neighbors) == 1:
+                    cut_graph[current_cut].add(neighbor_cut)
+                else:
+                    # Check for other neighbors
+                    for second_neighbor in (neighbors - {neighbor}):
+                        second_cut = cut_map[second_neighbor]
+                        if current_cut == second_cut:
+                            cut_graph[current_cut].add(neighbor_cut)
+                        # Ignore neighbors in the same cut
+                        if neighbor_cut == second_cut:
+                            cut_graph[current_cut].add(neighbor_cut)
+                            break
+                        # If a second neighbor belongs to a different cut, check for reachability and the non-reachable neighbor
+                        if self.is_reachable(neighbor, second_neighbor):
+                            cut_graph[current_cut].add(neighbor_cut)
+                            break
+
+        return Graph(cut_graph), cut_map
+
+    def all_pairs_reachability_dag(self) -> Dict[str, Set[str]]:
+        """
+        Compute reachability for all pairs in a DAG.
+        :return: Dictionary with reachability sets.
+        """
+        nodes = list(self.graph.keys())
+        reach = {node: set() for node in nodes}
+        
+        for node in nodes:
+            reach[node] = set(self.dfs(node))
+        
+        return reach
+
+    def find_unreachable_pairs(self) -> List[Tuple[str, str]]:
+        """
+        Find all pairs of nodes that are not reachable from each other.
+        :return: List of unreachable pairs.
+        """
+        reach = self.all_pairs_reachability_dag()
+        nodes = self.get_all_nodes()
+        non_reachable_pairs = set()
+        
+        for i in range(len(nodes)):
+            for j in range(i + 1, len(nodes)):
+                u = nodes[i]
+                v = nodes[j]
```

Comment:
Would be more straight forward to use enumerate() for that purpose instead of range(len())